hpx/plugins/parcelport/mpi/mpi_environment.hpp

```diff
@@ -29,27 +30,38 @@ namespace hpx { namespace util
         static int rank();
         static int size();
 
+        static MPI_Comm& communicator();
+
         static std::string get_processor_name();
 
+        struct scoped_lock
+        {
+            scoped_lock();
+            ~scoped_lock();
+        };
+
+        struct scoped_try_lock
+        {
+            scoped_try_lock();
+            ~scoped_try_lock();
+            bool locked;
+        };
+
+        typedef hpx::lcos::local::spinlock mutex_type;
+
     private:
-        static bool enabled_;
-        static bool has_called_init_;
-        static int provided_threading_flag_;
-    };
-}}
 
-#else
+        static mutex_type mtx_;
 
-namespace hpx { namespace util
-{
-    struct command_line_handling;
+        static void lock();
+        static bool try_lock();
+        static void unlock();
```

Comment:
The look is used to protect the calls to MPI. Locking gets automatically disabled if we have a multithreaded MPI at hand.